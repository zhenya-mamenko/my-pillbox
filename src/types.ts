export interface Pill {
  id?: number;
  name: string;
  description: string;
  bestBefore: string;
  imageData: string;
};
