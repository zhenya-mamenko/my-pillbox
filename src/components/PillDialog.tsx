import { ComponentProps, useEffect, useState, useRef, } from 'react';
import { Button, Pressable, Text, TextInput, StyleSheet, View, Dimensions, } from 'react-native';
import { DateTimePickerAndroid, AndroidNativeProps } from '@react-native-community/datetimepicker';
import { Image } from 'expo-image';
import { launchCameraAsync, type ImagePickerAsset} from 'expo-image-picker';
import { ImageManipulator } from 'expo-image-manipulator';
import SwipeModal, { SwipeModalPublicMethods } from '@birdwingo/react-native-swipe-modal';

import * as db from '@/helpers/sqlite';
import t from '@/helpers/localization';
import { Pill } from '@/helpers/types';
import useThemeColor from '@/hooks/useThemeColor';


interface Props extends ComponentProps<any> {
  pill: Pill;
  provideActions: (actions: Function[]) => void;
  onClose: () => void;
}

export default function PillDialog({ pill, onClose, provideActions }: Props) {

  const styles = StyleSheet.create({
    modalContent: {
      backgroundColor: useThemeColor('surface'),
      borderColor: useThemeColor('borderDim'),
      borderStyle: 'solid',
      borderWidth: 1,
      bottom: 0,
      height: '100%',
      position: 'absolute',
      width: '100%',
      zIndex: 1000,
    },
    titleContainer: {
      alignItems: 'center',
      borderBottomWidth: 1,
      borderColor: useThemeColor('borderDim'),
      borderStyle: 'solid',
      flexDirection: 'row',
      height: 80,
      justifyContent: 'space-between',
      paddingHorizontal: 20,
    },
    title: {
      color: useThemeColor('text'),
      fontSize: 18,
      fontWeight: 'bold',
    },
    photoImage: {
      borderRadius: 10,
      height: '35%',
      margin: 20,
      minHeight: 220,
      padding: 10,
    },
    input: {
      backgroundColor: useThemeColor('background'),
      borderColor: useThemeColor('borderDim'),
      borderRadius: 10,
      borderStyle: 'solid',
      borderWidth: 1,
      color: useThemeColor('text'),
      fontSize: 16,
      marginBottom: 20,
      marginHorizontal: 10,
      minHeight: 45,
      paddingHorizontal: 10,
      textAlignVertical: 'top',
      width: '100%',
    },
    inputsContainer: {
      alignItems: 'center',
      flex: 1,
      flexDirection: 'column',
      justifyContent: 'space-between',
      marginBottom: 20,
      paddingHorizontal: 10,
    },
  });

  const modalRef = useRef<SwipeModalPublicMethods>(null);

  const showModal = () => modalRef.current?.show();
  const hideModal = () => modalRef.current?.hide();
  provideActions([showModal, hideModal]);

  const [image, setImage] = useState("");
  const [name, setName] = useState("");
  const [description, setDescription] = useState("");
  const [dateBestBefore, setDateBestBefore] = useState(new Date());

  useEffect(() => {
    setImage(pill.imageData);
    setName(pill.name);
    setDescription(pill.description);
    setDateBestBefore(new Date(pill.bestBefore));
  }, [pill]);

  const pickImage = async () => {
    let result = await launchCameraAsync({
      allowsEditing: true,
      aspect: [1, 1],
      exif: false,
      quality: 0.85,
    });

    if (!result.canceled) {
      let imageURI = (result.assets[0] as ImagePickerAsset).uri;
      const context = ImageManipulator.manipulate(imageURI);
      context.resize({ width: 320, height: 320 });
      const imageRef = await context.renderAsync();
      imageURI = (await imageRef.saveAsync({ base64: true })).base64 as string;
      setImage(imageURI);
    }
  }

  const textDimColor = useThemeColor('textDim');
  const textColor = useThemeColor('text');
  const openDatePicker = () => {
    const params: AndroidNativeProps = {
      display: 'spinner',
      mode: 'date',
      negativeButton: {
        label: t('cancel'),
        textColor: textDimColor,
      },
      positiveButton: {
        label: t('select'),
        textColor: textColor,
      },
      value: dateBestBefore,
      onChange: (event, newDate) => setDateBestBefore(newDate ?? dateBestBefore),
    }
    DateTimePickerAndroid.open(params);
  }

  const SavePill = () => {
    const data = {
      name,
      description,
      bestBefore: dateBestBefore.toISOString().split('T')[0],
      imageData: image,
    }
    if (pill.id && pill.id !== 0) {
      db.updatePill({
        ...data,
        id: pill.id
      });
    } else {
      db.insertPill(data);
    }
    onClose();
  }

  return (
    <SwipeModal ref={modalRef}
      closeTrigger={'minHeight'}
      closeTriggerValue={500}
      hideKeyboardOnShow={false}
      maxHeight={Dimensions.get('screen').height * 0.75}
      showBar={false}
      useKeyboardAvoidingView={true}
      wrapInGestureHandlerRootView={true}
      onHide={onClose}
    >
      <View style={styles.modalContent}>
        <View style={styles.titleContainer}>
          <Text style={styles.title}>
            { (pill.id && pill.id !== 0) ? t('pill_edit') : t('pill_add') }
          </Text>
          <Button
            disabled={!image || !name}
            color={useThemeColor('primary')}
            title={`    ${t('save')}    `}
            onPress={SavePill}
          />
        </View>
        <Pressable onPress={pickImage} style={styles.photoImage}>
          <Image
            contentFit="contain"
            placeholder={require("../assets/images/stub.png")}
            source={{ uri: !image ? null : `data:image/jpg;base64,${image}` }}
            style={{ height: '100%', width: '100%' }}
          />
        </Pressable>
        <View style={styles.inputsContainer}>
          <TextInput style={styles.input}
            defaultValue={name}
            onChangeText={newName => setName(newName)}
            placeholder={ t('pill_name') }
            placeholderTextColor={ useThemeColor('textDim') }
          />
          <TextInput style={{ ...styles.input, height: '40%', }}
            defaultValue={description}
            multiline={true}
            onChangeText={newDescription => setDescription(newDescription)}
            placeholder={ t('pill_description') }
            placeholderTextColor={ useThemeColor('textDim') }
          />
          <Pressable
            style={{ padding: 0, width: '100%', }}
            onPress={openDatePicker}>
            <Text style={{ ...styles.input, height: 45, paddingVertical: 10, marginHorizontal: 0, }}>
              {`${t('pill_best_before')}: ${dateBestBefore.toLocaleDateString()}`}
            </Text>
          </Pressable>
        </View>
      </View>
    </SwipeModal>
  );
}
