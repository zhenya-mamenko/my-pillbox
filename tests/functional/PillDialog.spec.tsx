
import '@testing-library/react-native/extend-expect';
import { render, screen, userEvent, } from '@testing-library/react-native';
import PillDialog from '@/components/PillDialog';
import { Pill } from '@/types';


global.requestAnimationFrame = jest.fn();

jest.mock('@/hooks/useThemeColor', () => (color: string) => {
  const colors: any = {
    background: '#ffffff',
    border: '#000000',
    borderDim: '#ccccccc',
    primary: '#555555',
    surface: '#ffffff',
    text: '#000000',
    textDim: '#888888',
  }
  return colors[color];
});

jest.mock('@/helpers/localization', () => ({
  __esModule: true,
  default: (s: string) => s
}));

jest.mock('@birdwingo/react-native-swipe-modal', () => {
  const { forwardRef } = require('react');
  return {
    __esModule: true,
    // eslint-disable-next-line react/display-name
    default: forwardRef(({children}: {children: any}, ref: any) => <div>{children}</div>),
  };
});

const mockSQLite = {
  insertPill: jest.fn(),
  updatePill: jest.fn(),
}
jest.mock('@/helpers/sqlite', () => {
  return {
    __esModule: true,
    insertPill: (pill: Pill) => mockSQLite.insertPill(pill),
    updatePill: (pill: Pill) => mockSQLite.updatePill(pill),
  };
});

describe('PillDialog', () => {

  const pill = {
    id: 1,
    name: 'Pill 1',
    description: 'Description',
    bestBefore: '2025-10-01',
    imageData: 'iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII',
  }

  beforeAll(() => {
    jest.spyOn(Date.prototype, 'toLocaleDateString').mockImplementation(() => '2025-10-01');
  });

  it('Should render', () => {
    render(
      <PillDialog
        pill={pill}
        provideActions={() => {}}
        onClose={() => {}}
      />
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('Should contain right data', async () => {
    render(
      <PillDialog
        pill={pill}
        provideActions={() => {}}
        onClose={() => {}}
      />
    );

    expect(screen.getByTestId('dialog-title')).toHaveTextContent('pill_edit');
    expect(screen.getByTestId('dialog-name-input')).toHaveDisplayValue(pill.name);
    expect(screen.getByTestId('dialog-description-input')).toHaveDisplayValue(pill.description);
    expect(screen.getByTestId('dialog-best-before')).toHaveTextContent(`pill_best_before: ${(new Date(pill.bestBefore)).toLocaleDateString()}`);
    expect(screen.getByTestId('dialog-image')).toHaveProp('source', [{ uri: `data:image/jpg;base64,${pill.imageData}` }]);
  });

  it('Should call savePill', async () => {
    render(
      <PillDialog
        pill={pill}
        provideActions={() => {}}
        onClose={() => {}}
      />
    );

    const user = userEvent.setup();
    await user.press(screen.getByTestId('dialog-save-button'));

    expect(mockSQLite.updatePill).toHaveBeenCalledWith(pill);
  });

  it('Should call insertPill', async () => {
    const newPill: Pill = {...pill};
    delete newPill.id;
    newPill.name = '';
    newPill.description = '';
    render(
      <PillDialog
        pill={newPill}
        provideActions={() => {}}
        onClose={() => {}}
      />
    );

    expect(screen.getByTestId('dialog-save-button')).toBeDisabled();
    expect(screen.getByTestId('dialog-title')).toHaveTextContent('pill_add');

    const user = userEvent.setup();
    await user.type(screen.getByTestId('dialog-name-input'), 'Pill 2');
    await user.type(screen.getByTestId('dialog-description-input'), 'Description 2');

    expect(screen.getByTestId('dialog-save-button')).not.toBeDisabled();

    await user.press(screen.getByTestId('dialog-save-button'));

    expect(mockSQLite.insertPill).toHaveBeenCalledWith({ ...newPill, name: 'Pill 2', description: 'Description 2' });
  });

});
