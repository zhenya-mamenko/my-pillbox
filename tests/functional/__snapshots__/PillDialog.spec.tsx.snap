// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PillDialog Should render 1`] = `
<div>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#ccccccc",
        "borderStyle": "solid",
        "borderWidth": 1,
        "bottom": 0,
        "height": "100%",
        "position": "absolute",
        "width": "100%",
        "zIndex": 1000,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#ccccccc",
          "borderStyle": "solid",
          "flexDirection": "row",
          "height": 80,
          "justifyContent": "space-between",
          "paddingHorizontal": 20,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
        testID="dialog-title"
      >
        pill_edit
      </Text>
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="dialog-save-button"
      >
        <View
          style={
            [
              {},
            ]
          }
        >
          <Text
            disabled={false}
            style={
              [
                {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                {
                  "color": "#555555",
                },
              ]
            }
          >
                save    
          </Text>
        </View>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderRadius": 10,
          "height": "35%",
          "margin": 20,
          "minHeight": 220,
          "padding": 10,
        }
      }
      testID="dialog-image-button"
    >
      <ViewManagerAdapter_ExpoImage
        containerViewRef={"[React.ref]"}
        contentFit="contain"
        contentPosition={
          {
            "left": "50%",
            "top": "50%",
          }
        }
        height="100%"
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={
          [
            {
              "testUri": "../../../src/assets/images/stub.png",
            },
          ]
        }
        source={
          [
            {
              "uri": "data:image/jpg;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII",
            },
          ]
        }
        style={
          {
            "height": "100%",
            "width": "100%",
          }
        }
        testID="dialog-image"
        transition={null}
        width="100%"
      />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "space-between",
          "marginBottom": 20,
          "paddingHorizontal": 10,
        }
      }
    >
      <TextInput
        defaultValue="Pill 1"
        onChangeText={[Function]}
        placeholder="pill_name"
        placeholderTextColor="#888888"
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#ccccccc",
            "borderRadius": 10,
            "borderStyle": "solid",
            "borderWidth": 1,
            "color": "#000000",
            "fontSize": 16,
            "marginBottom": 20,
            "marginHorizontal": 10,
            "minHeight": 45,
            "paddingHorizontal": 10,
            "textAlignVertical": "top",
            "width": "100%",
          }
        }
        testID="dialog-name-input"
      />
      <TextInput
        defaultValue="Description"
        multiline={true}
        onChangeText={[Function]}
        placeholder="pill_description"
        placeholderTextColor="#888888"
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#ccccccc",
            "borderRadius": 10,
            "borderStyle": "solid",
            "borderWidth": 1,
            "color": "#000000",
            "fontSize": 16,
            "height": "40%",
            "marginBottom": 20,
            "marginHorizontal": 10,
            "minHeight": 45,
            "paddingHorizontal": 10,
            "textAlignVertical": "top",
            "width": "100%",
          }
        }
        testID="dialog-description-input"
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "padding": 0,
            "width": "100%",
          }
        }
        testID="dialog-best-before-input"
      >
        <Text
          style={
            {
              "backgroundColor": "#ffffff",
              "borderColor": "#ccccccc",
              "borderRadius": 10,
              "borderStyle": "solid",
              "borderWidth": 1,
              "color": "#000000",
              "fontSize": 16,
              "height": 45,
              "marginBottom": 20,
              "marginHorizontal": 0,
              "minHeight": 45,
              "paddingHorizontal": 10,
              "paddingVertical": 10,
              "textAlignVertical": "top",
              "width": "100%",
            }
          }
          testID="dialog-best-before"
        >
          pill_best_before: 01.10.2025
        </Text>
      </View>
    </View>
  </View>
</div>
`;
